pkg/agent/monitortool/latency_store.go

```diff
@@ -0,0 +1,263 @@
+	nodeIPLatencyMap map[string]*NodeIPLatencyEntry
+	// The map of Node name to Node IP, it will be changed by Node watcher
+	// If the agent is running in networkPolicyOnly mode, the value will be the transport IP of the Node.
+	// Otherwise, the value will be the gateway IP of the Node
+	nodeTargetIPsMap map[string][]net.IP
+}
+
+// NodeIPLatencyEntry is the entry of the latency map.
+type NodeIPLatencyEntry struct {
+	// The timestamp of the last sent packet
+	LastSendTime time.Time
+	// The timestamp of the last received packet
+	LastRecvTime time.Time
+	// The last valid rtt of the connection
+	LastMeasuredRTT time.Duration
+}
+
+// NewLatencyStore creates a new LatencyStore.
+func NewLatencyStore(isNetworkPolicyOnly bool) *LatencyStore {
+	store := &LatencyStore{
+		nodeIPLatencyMap:    make(map[string]*NodeIPLatencyEntry),
+		nodeTargetIPsMap:    make(map[string][]net.IP),
+		isNetworkPolicyOnly: isNetworkPolicyOnly,
+	}
+
+	return store
+}
+
+// getNodeIPLatencyEntry returns the NodeIPLatencyEntry for the given Node IP
+// For now, it is only used for testing purposes.
+func (l *LatencyStore) getNodeIPLatencyEntry(nodeIP string) (*NodeIPLatencyEntry, bool) {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+	entry, ok := l.nodeIPLatencyMap[nodeIP]
+	return entry, ok
+}
+
+// DeleteNodeIPLatencyEntry deletes the NodeIPLatencyEntry for the given Node IP
+func (l *LatencyStore) DeleteNodeIPLatencyEntry(nodeIP string) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	delete(l.nodeIPLatencyMap, nodeIP)
+}
+
+// SetNodeIPLatencyEntry sets the NodeIPLatencyEntry for the given Node IP
+func (l *LatencyStore) SetNodeIPLatencyEntry(nodeIP string, mutator func(entry *NodeIPLatencyEntry)) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	entry, ok := l.nodeIPLatencyMap[nodeIP]
+	if !ok {
+		entry = &NodeIPLatencyEntry{}
+		l.nodeIPLatencyMap[nodeIP] = entry
+	}
+
+	mutator(entry)
+}
+
+// ListLatencies returns the map of Node IP to latency entry
+func (l *LatencyStore) ListLatencies() map[string]*NodeIPLatencyEntry {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+
+	return l.nodeIPLatencyMap
+}
+
+// addNode adds a Node to the latency store
+func (l *LatencyStore) addNode(node *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	l.updateNodeMap(node)
+}
+
+// deleteNode deletes a Node from the latency store
+func (l *LatencyStore) deleteNode(node *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	nodeIPs, ok := l.nodeTargetIPsMap[node.Name]
+	if !ok {
+		klog.ErrorS(nil, "Failed to get IPs for Node", "nodeName", node.Name)
+		return
+	}
+
+	for _, ip := range nodeIPs {
+		delete(l.nodeIPLatencyMap, ip.String())
+	}
+
+	delete(l.nodeTargetIPsMap, node.Name)
+}
+
+// updateNode updates a Node in the latency store
+func (l *LatencyStore) updateNode(old *corev1.Node, new *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	// Delete from the node IP map
+	delete(l.nodeTargetIPsMap, old.Name)
```

Comment:
Yes, it may not be necessary to delete by node name, this was originally used to clear the old node name. but it looks like it should only trigger when the node is offline.

Reply:
There is no such thing as "old node name". A Node name doesn't change until the Node is deleted, and that's handled by `deleteNode`.
I think the Node's IP may change, but `l.updateNodeMap(new)` handles that correctly anyway.

Reply:
That's right, I have removed them.